hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/dml/TestInsertTable.scala

```diff
@@ -1668,47 +1668,102 @@ class TestInsertTable extends HoodieSparkSqlTestBase {
-                 | (7, 'a1,1', 10, 1000, "2021-01-05"),
-                 | (8, 'a1,1', 10, 1000, "2021-01-05"),
-                 | (10, 'a3,3', 30, 3000, "2021-01-05")
-               """.stripMargin)
-
-            checkAnswer(s"select count(distinct _hoodie_file_name) from $tableName where dt = '2021-01-05'")(
-              Seq(8)
-            )
+              // would generate 6 other files in partition(dt = '2021-01-05')
+              spark.sql(
+                s"""
+                   | insert into $tableName values
+                   | (4, 'a1,1', 10, 1000, "2021-01-05"),
+                   | (5, 'a1,1', 10, 1000, "2021-01-05"),
+                   | (6, 'a1,1', 10, 1000, "2021-01-05"),
+                   | (7, 'a1,1', 10, 1000, "2021-01-05"),
+                   | (8, 'a1,1', 10, 1000, "2021-01-05"),
+                   | (10, 'a3,3', 30, 3000, "2021-01-05")
+                 """.stripMargin)
+
+              checkAnswer(s"select count(distinct _hoodie_file_name) from $tableName where dt = '2021-01-05'")(
+                Seq(8)
+              )
+            }
           }
         }
       }
     }
   }
 
+  test("Test not supported multiple BULK INSERTs into SIMPLE BUCKET for COW without Spark native Row") {
+    withSQLConf("hoodie.datasource.write.operation" -> "bulk_insert",
+      "hoodie.bulkinsert.shuffle.parallelism" -> "1") {
+      withTempDir { tmp =>
+        val tableName = generateTableName
+        spark.sql(
+          s"""
+             |create table $tableName (
+             |  id long,
+             |  name string,
+             |  ts int,
+             |  par string
+             |) using hudi
+             | tblproperties (
+             | primaryKey = 'id,name',
+             | type = 'cow',
+             | preCombineField = 'ts',
+             | hoodie.index.type = 'BUCKET',
+             | hoodie.index.bucket.engine = 'SIMPLE',
+             | hoodie.bucket.index.num.buckets = '4',
+             | hoodie.bucket.index.hash.field = 'id,name',
+             | hoodie.datasource.write.row.writer.enable = 'false')
+             | partitioned by (par)
+             | location '${tmp.getCanonicalPath}'
+             """.stripMargin)
+        //   `id,name` -> `bucketId` if there are 4 buckets
+        //    5,'a1,1' ->    1
+        //    6,'a6,6' ->    2
+        //    9,'a3,3' ->    1
+        // 13,'a13,13' ->    2
+        //     24,'cd' ->    0
+        // buckets 1 & 2 into partition 'main', bucket 1 into partition 'side'
+        spark.sql(s"insert into $tableName values (5, 'a1,1', 1, 'main'), (6, 'a6,6', 1, 'main'), (9, 'a3,3', 1, 'side')")
+        // bucket 1 into 'main', bucket 2 into 'side', the whole insert will fail due to existed bucket 1 in 'main'
+        val causeRegex = "Multiple bulk insert.* COW.* not supported.*"
+        checkExceptionMatch(s"insert into $tableName values (9, 'a3,3', 2, 'main'), (13, 'a13,13', 1, 'side')")(causeRegex)
```

Comment:
@danny0405 , this test illustrates proposed by this MR behavior. Note that before these changes, we could write silently broken COW table consisted from base and log files.
Also, this exception is thrown only for hoodie.datasource.write.row.writer.enable = 'false', which is true by default.

Reply:
Why the spark native `Row` induces discrepancies here?

Reply:
It's because we have separate processing for this case, and I also added corresponding task https://issues.apache.org/jira/browse/HUDI-7757 previously to resolve this hidden branching in the middle of huge `HoodieSparkSqlWriter::writeInternal`:
https://github.com/apache/hudi/blob/f4e810b4db3d1c5d01ed24488fe603c86c3f2f36/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala#L488-L493
but didn't found quick solution then in the past.

Reply:
And in the case of `ENABLE_ROW_WRITER=true` we use `BaseDatasetBulkInsertCommitActionExecutor` instead of `SparkBulkInsertCommitActionExecutor`.

Reply:
I realized now, that we have limitations on parquet size, and when, for instance, `0001_fileGroup1` file is big enough, then we will continue to write to another `0001_fileGroup2` file.
So, maybe we could allow to generate new file groups in the case of multiple bulk inserts.

Reply:
Ok, I checked is it possible just to generate new file group for existed bucket, and found that there would be problems in `HoodieSimpleBucketIndex::loadBucketIdToFileIdMappingForPartition`:
https://github.com/apache/hudi/blob/f780c92d26dfe38c99cd1c1a9a71b00006e82c59/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/bucket/HoodieSimpleBucketIndex.java#L88-L91

Reply:
> And in the case of `ENABLE_ROW_WRITER = true` we use `BaseDatasetBulkInsertCommitActionExecutor` instead of `SparkBulkInsertCommitActionExecutor`.

As i see from documentation and @Danny's comment (https://github.com/apache/hudi/pull/12245#discussion_r1851871655), bulk_insert operation is "for initial loading/bootstrapping a Hudi table at first" (https://hudi.apache.org/docs/write_operations#bulk_insert), so it should be used only on empty tables. If you've already made one bulk_insert - the table is not empty anymore, so all subsequent operations must be upserts to avoid any potential issues with data consistency.
So my opinion is: for COW tables with Simple Bucket index we should throw exception on attempting to bulk_insert  to non-empty table, no matter what `ENABLE_ROW_WRITER` value is configured.

Reply:
@geserdugarov 
> So, maybe we could allow to generate new file groups in the case of multiple bulk inserts. @danny0405 , what do you think about it?

The reader can not work correctly if there are file groups that does not belong to the contracted buckets under bucket index. You can not generate more buckets than the configured number. I agree @wombatu-kun , let's throw exception here for COW table in native and non-native scenarios.

Reply:
@danny0405 , thanks for discussion. Then current changes in this MR could be merged (51f1bb1996c8c15d094dfe56f20004d8da0440a2 throws exception for COW when we try to do multiple bulk inserts). And I will try to find solution for native Spark row writer case, in this case we have completely different call stack.